src/test/java/com/iexec/worker/feign/LoginServiceTests.java

```diff
@@ -138,23 +141,60 @@ void shouldLogin() {
         );
     }
 
+    /**
+     * <ul>
+     *     <li>Spawn N threads
+     *     <li>Each thread tries to log in
+     *     <li>When the first thread enters the login method, it locks until N-1 threads have tried to log in
+     *     <li>In the end, a single thread should have retrieved a token
+     *     while the others should have had a "login already ongoing" message
+     * </ul>
+     */
     @Test
-    void shouldLoginOnceOnSimultaneousCalls(CapturedOutput output) throws InterruptedException, ExecutionException, TimeoutException {
-        Credentials credentials = generateCredentials();
+    void shouldLoginOnceOnSimultaneousCalls(CapturedOutput output)
+            throws InterruptedException,
+            ExecutionException,
+            TimeoutException {
+        // Pre-conditions
+        final int threadsCount = 3;
+        final ExecutorService executor = Executors.newFixedThreadPool(threadsCount);
+        final CountDownLatch gate = new CountDownLatch(threadsCount - 1);
+        final Answer<Object> waitForOtherThreads = invocation -> {
+            Awaitility.waitAtMost(1, TimeUnit.SECONDS)
+                    .pollDelay(5, TimeUnit.MILLISECONDS)
+                    .until(() -> gate.getCount() == 0);
+            return "token";
+        };
+        final Runnable loginAndDecrement = () -> {
+            loginService.login();
+            gate.countDown();
+        };
+
+        final Credentials credentials = generateCredentials();
         when(credentialsService.getCredentials()).thenReturn(credentials);
         when(coreClient.getChallenge(credentials.getAddress())).thenReturn("challenge");
-        Signature signature = SignatureUtils.hashAndSign("challenge", credentials.getAddress(), credentials.getEcKeyPair());
-        when(coreClient.login(credentials.getAddress(), signature)).thenReturn("token");
-        CompletableFuture<Void> run1 = CompletableFuture.runAsync(() -> loginService.login());
-        CompletableFuture<Void> run2 = CompletableFuture.runAsync(() -> loginService.login());
-        CompletableFuture<Void> run3 = CompletableFuture.runAsync(() -> loginService.login());
-        CompletableFuture.allOf(run1, run2, run3).get(1L, TimeUnit.SECONDS);
+        final Signature signature = SignatureUtils.hashAndSign("challenge", credentials.getAddress(), credentials.getEcKeyPair());
+        when(coreClient.login(credentials.getAddress(), signature))
+                .then(waitForOtherThreads)
+                .thenReturn("token");
+
+        // Execution
+        final CompletableFuture<?>[] completableFutures = IntStream.range(0, threadsCount)
```

Comment:
Is it not possible to keep `CompletableFuture<Void>` here ?

Reply:
We could, but that would lead to other issues. Either we have to cast the expression to `CompletableFuture<Void>[]` which throws a nice `unchecked cast` warning:
```java
        final CompletableFuture<Void>[] completableFutures = (CompletableFuture<Void>[]) IntStream.range(0, threadsCount)
                .mapToObj(i -> CompletableFuture.runAsync(loginAndDecrement, executor))
                .toArray(CompletableFuture<?>[]::new);
```
Either we try and instantiate the generic array from the `toArray` call:
```java
        final CompletableFuture<Void>[] completableFutures = IntStream.range(0, threadsCount)
                .mapToObj(i -> CompletableFuture.runAsync(loginAndDecrement, executor))
                .toArray(CompletableFuture<Void>[]::new);
```
In the latter options, the compiler doesn't let us create a generic array. I guess it's due on how generics are handled by Java's compiler.

This leads to me the following question: do we need to know which type these futures will return? I mean, we don't use the return value anyway. So IMHO there's no need to try and force their generic type.

Reply:
ok, I get it, we can keep `CompletableFuture<?>` with an erased type here.